lib/src/multires_stochastic_texture_synthesis.rs

```diff
@@ -1191,6 +1201,213 @@ impl PrerenderedU8Function {
+                closest_point_on_boundary_y: chunk_y * self.chunk_size as i32 - self.offset_y,
+            },
+            ChunkSearchInfo {
+                x: chunk_x - 1,
+                y: chunk_y - 1,
+                center: false,
+                closest_point_on_boundary_x: chunk_x * self.chunk_size as i32 - self.offset_x,
+                closest_point_on_boundary_y: chunk_y * self.chunk_size as i32 - self.offset_y,
+            },
+        ];
+        // Note locking all of them at different times is fine as opposed to
+        // all at once but this behavior should be noted
+        let mut tmp_result: Vec<(i32, i32, i32)> = Vec::with_capacity(k * 9);
+        result.clear();
+        result.reserve(k);
+
+        // this isn't really the kth best distance but should be good enough for how simple it is
+        let mut kth_best_squared_distance = i32::max_value();
+        for place_to_look in places_to_look.iter() {
+            if place_to_look.x >= 0
+                && place_to_look.x < self.grid_width as i32
+                && place_to_look.y >= 0
+                && place_to_look.y < self.grid_height as i32
+            {
+                let is_center = place_to_look.center;
+
+                // a tiny optimization to help us throw far away neighbors
+                // saves us a decent amount of reads
+                if !is_center {
+                    let squared_distance_to_closest_possible_point_on_chunk = (x as i32
+                        - place_to_look.closest_point_on_boundary_x)
+                        * (x as i32 - place_to_look.closest_point_on_boundary_x)
+                        + (y as i32 - place_to_look.closest_point_on_boundary_y)
+                            * (y as i32 - place_to_look.closest_point_on_boundary_y);
+
+                    if squared_distance_to_closest_possible_point_on_chunk
+                        > kth_best_squared_distance
+                    {
+                        continue;
+                    }
+                }
+
+                let my_tree_index = self.get_space(place_to_look.x as u32, place_to_look.y as u32);
+                let my_rtree = &self.rtrees[my_tree_index];
+                tmp_result.extend(
+                    my_rtree
+                        .read()
+                        .unwrap()
+                        .nearest_neighbor_iter(&[x as i32, y as i32])
+                        .take(k)
+                        .map(|a| {
+                            (
+                                (*a)[0],
+                                (*a)[1],
+                                ((*a)[0] - x as i32) * ((*a)[0] - x as i32)
+                                    + ((*a)[1] - y as i32) * ((*a)[1] - y as i32),
+                            )
+                        }),
+                );
+
+                // this isn't really the kth best distance but it's an okay approximation
```

Comment:
I think a good deal of "this isn't really X" could be avoided by just calling the variable differently. How about something like `kth_nearest_distance_upper_bound`?

Reply:
👍